File: 20190917/20190924/724. 寻找数组的中心索引.cc
/*
 * Description: 寻找数组的中心索引
 * Author: AskeyNil
 * CreateDate: 2019-09-24 11:32:28
 * LastEditors: AskeyNil
 *
 * *********************************
 * **                             **
 * **     　  你只有足够努力     　  **
 * **       才能看上去毫不费力       **
 * **                             **
 * *********************************
 *
 */

#include <iostream>
#include <numeric>
#include <vector>

using namespace std;

class Solution {
  public:
    int pivotIndex(vector<int> &nums) {
        int length = nums.size(), *left = nullptr, *right = nullptr;
        int cur_index = -1;
        for (int index = 0; index < length; index++) {
            if (left == nullptr) {
                left = new int(0);
                right = new int(accumulate(nums.begin() + 1, nums.end(), 0));
            } else {
                *left += nums[index - 1];
                *right -= nums[index];
            }
            if (left == right) {
                cur_index = index;
                break;
            }
        }
        delete left, delete right;
        return cur_index;
    }
};

File: 20190917/20190924/724. 寻找数组的中心索引.py
'''
Description: 寻找数组的中心索引
Author: AskeyNil
CreateDate: 2019-09-24 11:32:18
LastEditors: AskeyNil

*********************************
**                             **
**     　  你只有足够努力     　  **
**       才能看上去毫不费力       **
**                             **
*********************************

'''


class Solution:
    def pivotIndex(self, nums: [int]) -> int:
        length = len(nums)
        left = None
        right = None
        for index in range(length):
            if left == None:
                left = sum(nums[0:index])
                right = sum(nums[index+1:])
            else:
                left += nums[index-1]
                right -= nums[index]
            if left == right:
                return index + 1
        return -1
